src/software/ai/navigator/path_planner/bang_bang_trajectory_2d.cpp

```diff
@@ -0,0 +1,116 @@
+#include "software/ai/navigator/path_planner/bang_bang_trajectory_2d.h"
+
+BangBangTrajectory2D::BangBangTrajectory2D(const Point& initial_pos,
+                                           const Point& final_pos,
+                                           const Vector& initial_vel, double max_vel,
+                                           double max_accel, double max_decel)
+{
+    generate(initial_pos, final_pos, initial_vel, max_vel, max_accel, max_decel);
+}
+
+void BangBangTrajectory2D::generate(const Point& initial_pos, const Point& final_pos,
+                                    const Vector& initial_vel, double max_vel,
+                                    double max_accel, double max_decel)
+{
+    /**
+     * The following implementation is based on section IV of the paper
+     * "Trajectory generation for four wheeled omnidirectional vehicles"
+     *
+     * A 2D trajectories is represented as 2x 1D trajectories (one for x and one for y)
+     * which take the same amount of time to reach the destination.
+     *
+     * The max velocity, acceleration, and deceleration are assumed to be the kinematic
+     * constraints in 2D space, as opposed to the constraints for individual x and y
+     * components. To find the components of these constraints for each axis which will
+     * give us two trajectories with (almost) equal durations, we will use binary search
+     * over the components of the given kinematic constraints.
+     *
+     * We will start with alpha = PI/4 (45 deg) which will have the x and y components
+     * equivalent to each other. Depending on which 1D trajectory takes longer to reach
+     * the destination, we will either increase or decrease alpha by increment
+     * (= PI/(8*2^i) where i is the iteration of binary search we're at). This will
+     * result in the component which is taking longer to reach the destination to get a
+     * larger component of the kinematic constraints relative to the other component which
+     * reaches the destination faster.
+     */
+
+    double alpha     = M_PI / 4.0;
+    double increment = M_PI / 8.0;
+
+    while (increment > 1e-3)
+    {
+        const double cos = std::cos(alpha);
+        const double sin = std::sin(alpha);
+
+        x_trajectory.generate(initial_pos.x(), final_pos.x(), initial_vel.x(),
+                              max_vel * cos, max_accel * cos, max_decel * cos);
+        y_trajectory.generate(initial_pos.y(), final_pos.y(), initial_vel.y(),
+                              max_vel * sin, max_accel * sin, max_decel * sin);
+
```

Comment:
Is it worth improving the handling the case of a high initial velocity?

If we notice that the initial velocity for a component starts larger than the component's maximum allowed velocity, then we should try to make sure that the other component doesn't exceed the maximum global velocity for the duration it takes.

Reply:
I think it might be a bit difficult to handle this edge case as I think it would require further coupling between the two components (e.g. x is above its max velocity and is decelerating towards its max velocity, but y has to wait for x to reach its max velocity before being able to safely go to its own max velocity). Personally Im not sure if it's something that we need to handle.

Reply:
If its a problem, we'll handle it in the future

Reply:
Ya I think if it becomes a noticeable problem we can look into a more sophisticated approach. I don't think we need to overcomplicate now before doing in depth testing of it